File: cliente_automatico.py
# cliente_automatico.py
import zmq
from datetime import datetime
import time
import random
import string
import msgpack 

# --- Configuração ---
context = zmq.Context()
socket = context.socket(zmq.REQ)
socket.connect("tcp://broker:5557") # Conexão já estava correta

# --- Funções Auxiliares ---
def random_string(length=8):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(length))

# --- Lógica Principal ---

# 1. Login com usuário aleatório
user_name = f"bot_{random_string(5)}"
login_req = {
    "service": "login",
    "data": {"user": user_name, "timestamp": datetime.now().isoformat()}
}

# --- CORREÇÃO 1: Usar msgpack ---
print(f"Bot '{user_name}' tentando logar...")
socket.send(msgpack.packb(login_req, default=str))
reply_packed = socket.recv()
reply = msgpack.unpackb(reply_packed, raw=False)
# --- Fim da Correção 1 ---

if reply["data"]["status"] != "sucesso":
    print(f"Bot {user_name} falhou ao logar. Encerrando.")
    exit()
print(f"Bot '{user_name}' logado com sucesso.")

# Mensagens de exemplo para enviar
MENSAGENS = [
    "Olá a todos!", "Alguém aí?", "Este é um teste do sistema de mensagens.",
    "Que dia para programar!", "ZeroMQ é muito interessante.",
    "Testando, 1, 2, 3...", "Docker facilita muito a vida.",
    "Python é uma ótima linguagem.", "Quem quer café?", "Preciso de férias."
]

# 2. Loop infinito de publicação
while True:
    try:
        # Pega a lista de canais
        # --- CORREÇÃO 2: Usar msgpack ---
        socket.send(msgpack.packb({"service": "channels", "data": {"timestamp": datetime.now().isoformat()}}, default=str))
        channels_reply_packed = socket.recv()
        channels_reply = msgpack.unpackb(channels_reply_packed, raw=False)
        # --- Fim da Correção 2 ---
        
        available_channels = channels_reply.get("data", {}).get("channels", [])

        if not available_channels:
            # Se não houver canais, cria um
            new_channel = f"canal_{random_string(4)}"
            print(f"Bot '{user_name}' criando canal '{new_channel}'...")
            
            # --- CORREÇÃO 3: Usar msgpack ---
            socket.send(msgpack.packb({"service": "channel", "data": {"channel": new_channel, "timestamp": datetime.now().isoformat()}}, default=str))
            socket.recv() # Apenas consome a resposta (não precisamos desempacotar)
            # --- Fim da Correção 3 ---
            
            available_channels.append(new_channel)
        
        # Escolhe um canal aleatório
        target_channel = random.choice(available_channels)

        # Envia 10 mensagens
        print(f"Bot '{user_name}' vai enviar 10 mensagens para o canal '{target_channel}'.")
        for i in range(10):
            message_to_send = random.choice(MENSAGENS)
            publish_req = {
                "service": "publish",
                "data": {
                    "user": user_name,
                    "channel": target_channel,
                    "message": f"({i+1}/10) {message_to_send}",
                    "timestamp": datetime.now().isoformat()
                }
            }
            
            # --- CORREÇÃO 4: Usar msgpack ---
            socket.send(msgpack.packb(publish_req, default=str))
            socket.recv() # Consome a resposta
            # --- Fim da Correção 4 ---
            
            time.sleep(random.uniform(0.5, 2.0)) # Espera um pouco entre as mensagens

    except Exception as e:
        print(f"Bot '{user_name}' encontrou um erro: {e}")
        time.sleep(5) # Espera antes de tentar novamente

File: cliente_automatico.js
// cliente_automatico.js

const zmq = require("zeromq");
const msgpack = require("@msgpack/msgpack");

let logicalClock = 0;

// --- Configuração ---
// No Node.js, sockets são assíncronos
const socket = new zmq.Request();
socket.connect("tcp://broker:5557");
console.log("Cliente automático (JS) conectando ao tcp://broker:5557...");

// --- Funções Auxiliares ---
function randomString(length = 8) {
  const letters = 'abcdefghijklmnopqrstuvwxyz';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += letters.charAt(Math.floor(Math.random() * letters.length));
  }
  return result;
}

// Equivalente ao time.sleep() do Python
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Mensagens de exemplo
const MENSAGENS = [
    "Olá do Node.js!", "Alguém aí?", "Este é um teste do sistema de mensagens.",
    "Que dia para programar!", "ZeroMQ é muito interessante.",
    "Testando, 1, 2, 3...", "Docker facilita muito a vida.",
    "JavaScript também é legal.", "Quem quer café?", "Preciso de férias."
];

// --- Lógica Principal (Async) ---
// Precisamos de uma função 'async' para usar 'await'
async function main() {
  // 1. Login com usuário aleatório
  const userName = `bot_js_${randomString(5)}`;
  logicalClock++;
  const loginReq = {
    "service": "login",
    "data": {
      "user": userName, 
      "timestamp": new Date().toISOString(),
      "clock": logicalClock
    } // ISO format é o mesmo
  };

  try {
    console.log(`Bot '${userName}' tentando logar...(Clock: ${logicalClock})`);
    
    await socket.send(msgpack.encode(loginReq));
    const [replyPacked] = await socket.receive(); 
    const reply = msgpack.decode(replyPacked); 

    const receivedClock = reply.data?.clock || 0;
    logicalClock = Math.max(logicalClock, receivedClock);

    if (reply.data.status !== "sucesso") {
      console.error(`Bot ${userName} falhou ao logar. Encerrando.`);
      process.exit(1); // Encerra o script
    }
    console.log(`Bot '${userName}' logado com sucesso.`);

  } catch (err) {
    console.error("Erro no login:", err);
    process.exit(1);
  }

  // 2. Loop infinito de publicação
  while (true) {
    try {
      // Pega a lista de canais
      logicalClock++;
      const channelsReq = {
        "service": "channels", 
        "data": {
          "timestamp": new Date().toISOString(),
          "clock": logicalClock
        }
      };
      await socket.send(msgpack.encode(channelsReq));
      const [channelsReplyPacked] = await socket.receive();
      const channelsReply = msgpack.decode(channelsReplyPacked);
      const channelsClock = channelsReply.data?.clock || 0;
      logicalClock = Math.max(logicalClock, channelsClock);
      
      // 'data' pode não existir, 'channels' pode não existir.
      let availableChannels = channelsReply.data?.channels || []; 

      if (availableChannels.length === 0) {
        // Se não houver canais, cria um
        const newChannel = `canal_js_${randomString(4)}`;
        console.log(`Bot '${userName}' criando canal '${newChannel}'...`);
        logicalClock++;
        const createChannelReq = {
          "service": "channel", 
          "data": {
            "channel": newChannel, 
            "timestamp": new Date().toISOString(),
            "clock": logicalClock
          }
        };
        
        await socket.send(msgpack.encode(createChannelReq));

        const [createReplyPacked] = await socket.receive();
        const createReply = msgpack.decode(createReplyPacked);

        // Regra 2
        const createClock = createReply.data?.clock || 0;
        logicalClock = Math.max(logicalClock, createClock);
        availableChannels.push(newChannel);
      }

      // Escolhe um canal aleatório
      const targetChannel = availableChannels[Math.floor(Math.random() * availableChannels.length)];

      // Envia 10 mensagens
      console.log(`Bot '${userName}' vai enviar 10 mensagens para o canal '${targetChannel}'.`);
      for (let i = 0; i < 10; i++) {
        const messageToSend = MENSAGENS[Math.floor(Math.random() * MENSAGENS.length)];
        logicalClock++;
        const publishReq = {
          "service": "publish",
          "data": {
            "user": userName,
            "channel": targetChannel,
            "message": `(JS ${i+1}/10) ${messageToSend}`,
            "timestamp": new Date().toISOString(),
            "clock": logicalClock
          }
        };

        await socket.send(msgpack.encode(publishReq));
        const [pubReplyPacked] = await socket.receive();
        const pubReply = msgpack.decode(pubReplyPacked);

        const pubClock = pubReply.data?.clock || 0;
        logicalClock = Math.max(logicalClock, pubClock);
        
        // Espera entre 0.5s e 2.0s
        const waitTime = Math.random() * (2000 - 500) + 500; 
        await sleep(waitTime);
      }

    } catch (err) {
      console.error(`Bot '${userName}' encontrou um erro:`, err);
      await sleep(5000); // Espera antes de tentar novamente
    }
  }
}

// Inicia a função principal
main();